pkg/link/spoofcheck.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2021 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package link
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/networkplumbing/go-nft/nft"
+	"github.com/networkplumbing/go-nft/nft/schema"
+)
+
+const (
+	natTableName            = "nat"
+	preRoutingBaseChainName = "PREROUTING"
+)
+
+type applyAction func(*nft.Config) error
+type readAction func() (*nft.Config, error)
+
+type SpoofChecker struct {
+	iface      string
+	macAddress string
+	refID      string
+	applyFunc  applyAction
+	readFunc   readAction
+}
+
+func NewSpoofChecker(iface, macAddress, refID string) *SpoofChecker {
+	return &SpoofChecker{iface, macAddress, refID, nft.ApplyConfig, nft.ReadConfig}
+}
+
+func NewSpoofCheckerWithCustomActions(iface, macAddress, refID string, applyF applyAction, readF readAction) *SpoofChecker {
+	sc := NewSpoofChecker(iface, macAddress, refID)
+	sc.applyFunc = applyF
+	sc.readFunc = readF
+	return sc
+}
+
+// Setup applies nftables configuration to restrict traffic
+// from the provided interface. Only traffic with the mentioned mac address
+// is allowed to pass, all others are blocked.
+// The configuration follows the format libvirt and ebtables implemented, allowing
+// extensions to the rules in the future.
+// refID is used to label the rules with a unique comment, identifying the rule-set.
+//
+// In order to take advantage of the nftables configuration change atomicity, the
+// following steps are taken to apply the configuration:
+// - Declare the table and chains (they will be created in case not present).
+// - Apply the rules, while first flushing the iface/mac specific regular chain rules.
+// Note: Although two transactions are taken place, only the 2nd one where the rules
+// are added has a real impact on the system.
+func (sc *SpoofChecker) Setup() error {
```

Comment:
Because there is a need to flush first existing rules if they are there.
We cannot do it in one transaction because if the chain/table is not there yet, the flush operation will fail (failing to find the mentioned table/chain).

So the table/chain is declared first, if it is there, nothing will happen, if it is missing it will be created.
Then we can flush (knowing that it is there for sure).

Reply:
Can we add a comment with this?

Reply:
Ok, will try to add it to the func doc.

Reply:
Done